packages/orm/src/EntityManager.ts

```diff
@@ -612,67 +630,62 @@ export class EntityManager<C = {}> {
-              // will kick in the dataloader auto-batching and end up being smartly populated (granted via 1 query per
-              // entity type per "level" of resolution, instead of 1 single giant SQL query that inner joins everything
-              // in).
-              if (typeof hint === "string") {
-                return (entity as any)[hint].load(opts);
-              } else if (Array.isArray(hint)) {
-                return (hint as string[]).map((key) => (entity as any)[key].load(opts));
-              } else if (typeof hint === "object") {
-                return Object.entries(hint as object).map(([key, nestedHint]) => {
-                  const relation = (entity as any)[key];
-                  return relation.load(opts).then((result: any) => {
-                    if (
-                      Object.keys(nestedHint).length > 0 &&
-                      ((result instanceof Array && result.length > 0) || result !== undefined)
-                    ) {
-                      return this.populate(result, { hint: nestedHint, ...opts });
-                    }
-                    return result;
-                  });
-                });
-              } else {
-                throw new Error(`Unexpected hint ${hint}`);
-              }
+          const hints = Object.entries(normalizeHint(hintOpt as any));
+
+          // One breadth-width pass to ensure each relation is loaded
+          const loadPromises = list.flatMap((entity) => {
+            return hints.map(([key]) => {
+              const relation = (entity as any)[key];
+              return relation.isLoaded && !opts.forceReload ? undefined : (relation.load(opts) as Promise<any>);
             });
-          return Promise.all(promises);
+          });
+
+          // 2nd breadth-width pass to do nested load hints
+          return Promise.all(loadPromises).then(() => {
+            const nestedLoadPromises = hints.map(([key, nestedHint]) => {
+              if (Object.keys(nestedHint).length === 0) return;
+              // Unique for good measure?...
+              const results = [...new Set(list.map((entity) => toArray(entity[key].get)).flat())];
+              if (results.length === 0) return;
+              return this.populate(results, { hint: nestedHint, ...opts });
+            });
+            // After the nested hints are done, echo back the original now-loaded list
+            return Promise.all(nestedLoadPromises).then(() => list);
+          });
         },
+
         // We disable caching if `forceReload` is enabled to ensure we re-touch any previously-loaded
         // collections; otherwise using `cache: false` all the time does reduce the effectiveness of
         // the fan-in and results in duplicate populate calls.
         { cache: !opts.forceReload },
       ),
     );
 
-    // Purposefully use `then` instead of `async` as an optimization
-    return loader.loadMany(list).then(() => (!fn ? (entityOrList as any) : fn(entityOrList as any)));
```

Comment:
This `loadMany()` not checking its return value is the original bug that started this rewrite of `populate`.

What would happen is that `.load()` on an async property would blow up, and we would ignore it, but then since the promise was rejected, it would ~early return (i.e. not `allSettled` behavior), and code that used the `populate` results would start calling `.get` into not-yet-loaded relations, and get "not loaded" errors.